riscv-runtime/src/hash.rs

```diff
@@ -27,3 +29,69 @@ pub fn poseidon_gl(data: [u64; 12]) -> [u64; 4] {
 
     poseidon_gl_unsafe(data)
 }
+
+/// Calls the keccakf machine
+/// Return value is placed in the output array.
+pub fn keccakf(input: *const [u64; 25], output: *mut [u64; 25]) {
+    // TODO: uncomment this chunk after syscall implemented
+    // unsafe {
+        // // syscall inputs: memory pointer to input array and memory pointer to output array
+        // asm!("ecall", in("a0") input, in("a1") output, in("t0") u32::from(Syscall::KeccakF));
+    // }
+
+    // TODO: delete the following testing only chunk which uses tiny_keccak once syscall implemented
+    unsafe {
+        // Convert the input pointer to a mutable reference
+        let input_slice = &mut *(input as *mut [u64; 25]);
+
+        // Perform the keccakf operation in place
+        tiny_keccak_keccakf(input_slice);
+
+        // Copy the result to the output
+        let output_slice = &mut *output;
+        output_slice.copy_from_slice(input_slice);
+    }
+}
+
+// Output number of bytes for keccak-256 (32 bytes)
+const W: usize = 32;
+
+/// Keccak function that calls the keccakf machine
+/// Input is a byte array of arbitrary length and a delimiter byte
+/// Output is a byte array of length W
+pub fn keccak(data: &[u8], delim: u8) -> [u8; W] {
+    let mut b_toggle = [[0u8; 200]; 2];
```

Comment:
I think this can be simpler (no need for the toggle)
```
    let mut b = [[0u8; 200]; 2];
    let [mut b_input, mut b_output] = &mut b;
...
    // then, to swap:
    core::mem::swap(&mut b_input, &mut b_output);
```

Reply:
then, if keccakf takes normal references (& / &mut) I think you can just do:
`keccakf(mem::transmute(&mut b_input), mem::transmute(&mut b_output));`